include/experimental/__p0009_bits/utility.hpp

```diff
@@ -165,6 +165,63 @@ constexpr const auto& get(const tuple<Args...>& vals) { return vals.template get
 template<class ... Elements>
 tuple(Elements ...) -> tuple<Elements...>;
 #endif
+
+#if MDSPAN_HAS_CXX_17
+// std::in_range and friends, tagged for device execution
+// Backport from https://en.cppreference.com/w/cpp/utility/intcmp
+// and https://en.cppreference.com/w/cpp/utility/in_range
+template <class T, class U>
+MDSPAN_INLINE_FUNCTION constexpr bool cmp_less(T t, U u) noexcept {
+  if constexpr (std::is_signed_v<T> == std::is_signed_v<U>)
+    return t < u;
+  else if constexpr (std::is_signed_v<T>)
+    return t < 0 || std::make_unsigned_t<T>(t) < u;
+  else
+    return u >= 0 && t < std::make_unsigned_t<U>(u);
+}
+
+template <class T, class U>
+MDSPAN_INLINE_FUNCTION constexpr bool cmp_less_equal(T t, U u) noexcept {
+  return !cmp_less(u, t);
+}
+
+template <class T, class U>
+MDSPAN_INLINE_FUNCTION constexpr bool cmp_greater_equal(T t, U u) noexcept {
+  return !cmp_less(t, u);
+}
+
+template <class R, class T>
+MDSPAN_INLINE_FUNCTION constexpr bool in_range(T t) noexcept {
+  return cmp_greater_equal(t, std::numeric_limits<R>::min()) &&
+          cmp_less_equal(t, std::numeric_limits<R>::max());
+}
+
+template <typename T >
+MDSPAN_INLINE_FUNCTION constexpr bool
+check_mul_result_is_representable(T a, T b) {
+  if (b == 0 || a == 0)
+    return true;
+
+  // FIXME NVCC 11 Separate branch for this or nvcc complains about comparison of unsigned to 0
+  if constexpr (!std::is_signed_v<T>) {
+    return a <= std::numeric_limits<T>::max() / b;
+  } else {
+    // check overflow for positive a, b
+    if (b > 0 && a > std::numeric_limits<T>::max() / b)
+      return false;
+    // check overflow for negative a, b
+    if (b < 0 && a < std::numeric_limits<T>::max() / b)
+      return false;
+    // check underflow for positive b, negative a
+    if (b > 0 && a < std::numeric_limits<T>::min() / b)
+      return false;
+    // check underflow for negative b, positive a
+    if (b < 0 && a > std::numeric_limits<T>::min() / b)
+      return false;
+  }
```

Comment:
This is a function which is only used with extents right now. Let's not pay for the overhead of all these other checks for negative number combos.

```suggestion
  if constexpr (std::is_signed_v<T>) {
    if ( a < 0 || b < 0 ) return false;
  }
  return a <= std::numeric_limits<T>::max() / b;
```

Reply:
possibly could rename function to indicate we expect positive.